debounce.go

```diff
@@ -0,0 +1,105 @@
+// Pacage debounce provides a number of debounce patterns as a set of easy to
+// use functions.
+package debounce
+
+import (
+	"sync"
+	"time"
+)
+
+// New returns a debounced function that delays invoking f until after wait time
+// has elapsed since the last time the debounced function was invoked.
+//
+// The returned cancel function can be used to cancel any pending invocation of
+// f.
+//
+// Both debounced and cancel functions are safe for concurrent use in
+// goroutines, and can both be called multiple times.
+//
+// The debounced function does not wait for f to complete, so f needs to be
+// thread-safe as it may be invoked again before the previous invocation
+// completes.
+func New(wait time.Duration, f func()) (debounced func(), cancel func()) {
+	var mux sync.Mutex
+	timer := stoppedTimer(f)
+
+	debounced = func() {
+		mux.Lock()
+		defer mux.Unlock()
+
+		timer.Reset(wait)
+	}
+
+	cancel = func() {
+		mux.Lock()
+		defer mux.Unlock()
+
+		timer.Stop()
+	}
+
+	return debounced, cancel
+}
+
+// NewWithMaxWait returns a debounced function like New, but with a maximum wait
+// time of maxWait, which is the maximum time f is allowed to be delayed before
+// it is invoked.
+//
+// The returned cancel function can be used to cancel any pending invocation of
+// f.
+//
+// Both debounced and cancel functions are safe for concurrent use in
+// goroutines, and can both be called multiple times.
+//
+// The debounced function does not wait for f to complete, so f needs to be
+// thread-safe as it may be invoked again before the previous invocation
+// completes.
+func NewWithMaxWait(
+	wait, maxWait time.Duration,
+	f func(),
+) (debounced func(), cancel func()) {
+	var mux sync.Mutex
+	var dirty bool
+	var timer *time.Timer
+	var maxTimer *time.Timer
+
+	cb := func() {
+		mux.Lock()
+		defer mux.Unlock()
+
+		if !dirty {
+			return
+		}
+
+		go f()
```

Comment:
`New()` is sync while `NewWithMaxWait` seems to be async, but there is no mention of this difference.

Reply:
As far as I've understood, `New()` is async too, cause `time.AfterFunc()` runs the function in its own goroutine. I decided to spawn `fn` further out to its own goroutine here cause we're also calling `Stop()` on the timers, and didn't want the stops to slow down `fn`, or vice versa.